### PR TITLE
[MM-41996] Channel Members RHS

### PR DESCRIPTION
**Pull request by JulienTant**
#### Summary
This PR adds a new Channel members RHS, and an integration to it from the Channel Info RHS.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-41996
https://mattermost.atlassian.net/browse/MM-42973

#### Related Pull Requests
N/A

#### Screenshots

Screen Shot 2022-03-30 at 2 53 32 PM

Screen Shot 2022-03-30 at 2 54 38 PM

image

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Hey @JulienTant, nice work on this. It's awesome. A few things from my initial review:

1. If you are the only member of the channel, we should probably hide the 'Manage' button since it doesn't do anything in this case anyway.
2. While in 'Managing mode', can we make the `ESC` key cancel?
3. Also, I think we should change 'Cancel' to 'Done' while in Managing mode and make it a Primary button. I've updated the design file to reflect this.
image

4. In managing mode, the dropdown should have a hover and active state on it.
image

5. The icon size on the 'Add' button looks a bit too small. The text is right, but the icon should be `font-size: 14.4px;`
6. For the hover state of the 'send message' icon button, can we add a tooltip that says "Send message". Also, we need to add rounded corners (4px border-radius) to this icon button.
7. Can we adjust the padding on the main content of the RHS so that the 'Channel Admins' and the users listed align with the "X members" title?
<table>
<tr>
<td>Misaligned</td>
<td>Aligned</td>
</tr>
<tr>
<td>
image
</td>
<td>
image
</td>
</tr>
</table>
8. In the channel header, the members icon still opens a popover. We should make that open the members RHS now.

**Comment by JulienTant** [comment]
# Todo list:

- [x] If you are the only member of the channel, we should probably hide the 'Manage' button since it doesn't do anything in this case anyway.
- [x] While in 'Managing mode', can we make the ESC key cancel?
- Also, I think we should change 'Cancel' to 'Done' while in Managing mode and make it a Primary button. I've updated the design file to reflect this.
  - [x] Done
  - [x] Primary
- [ ] In managing mode, the dropdown should have a hover and active state on it.
- [x] The icon size on the 'Add' button looks a bit too small. The text is right, but the icon should be font-size: 14.4px;
- For the hover state of the 'send message' icon button, can we add a tooltip that says "Send message". Also, we need to add rounded corners (4px border-radius) to this icon button.
  - [ ] Hover message
  - [x] rounded corner
- [ ] Can we adjust the padding on the main content of the RHS so that the 'Channel Admins' and the users listed align with the "X members" title?
- [ ] In the channel header, the members icon still opens a popover. We should make that open the members RHS now.

**Review by matthewbirtch — APPROVED** [review]
Nicely done @JulienTant! It's looking great.

**Comment by neallred** [comment]
I think we shouldn't show the members info rhs if user is viewing a channel/group message and switches to a direct message. What makes sense to me is:

* If they have a previousRHSState where they were viewing a different RHS tab, move them to that.
* If there is no previousRHSState, close the RHS.

https://user-images.githubusercontent.com/13738432/161770727-96edd50c-571e-47a1-8849-82d4596b3b2d.mp4

File: components/channel_info_rhs/index.ts [hunk 1]
```diff
@@ -29,12 +29,19 @@ import {Channel} from 'mattermost-redux/types/channels';
 
 import RHS, {Props} from './channel_info_rhs';
 
+const EMPTY_CHANNEL_STATS = {
```
Comment (neallred) [inline components/channel_info_rhs/index.ts:32]: +1 for constant empty object.

File: components/channel_members_rhs/member.tsx [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    align-items: center;
+    padding: 8px 16px;
+    border-radius: 4px;
+
+    &:hover {
+        background: rgba(var(--center-channel-text-rgb), 0.08);
+        color: rgba(var(--center-channel-text-rgb), 0.56);
+        ${() => {
+        return css`
+            ${SendMessage} {
```
Comment (neallred) [inline components/channel_members_rhs/member.tsx:173]: Nice trick :)

File: components/channel_members_rhs/member_list.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    font-weight: 600;
+    font-size: 12px;
+    line-height: 28px;
+    letter-spacing: 0.02em;
+    text-transform: uppercase;
+    padding: 0px 12px;
+    color: rgba(var(--center-channel-color-rgb), 0.56);
+`;
+
+const Members = styled.div`
```
Comment (neallred) [inline components/channel_members_rhs/member_list.tsx:24]: Nit: remove since it does not have any styles.

File: e2e/cypress/tests/integration/channel/channel_members_rhs_spec.js [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    cy.visit(`/${testTeam.name}/channels/${testChannel.name}`);
+
+    // # Click on the channel info button
+    cy.get('#channel-info-btn').click();
+
+    // # Click on the Members menu
+    cy.uiGetRHS().findByText('Members').should('be.visible').click();
+}
+
+describe('Channel members RHS', () => {
```
Comment (neallred) [inline e2e/cypress/tests/integration/channel/channel_members_rhs_spec.js:24]: Nice tests!

File: reducers/views/search.ts [hunk 5]
```diff
@@ -23,6 +23,19 @@ function modalSearch(state = '', action: GenericAction) {
     }
 }
 
+function channelMembersRhsSearch(state = '', action: GenericAction) {
+    switch (action.type) {
+    case SearchTypes.SET_CHANNEL_MEMBERS_RHS_SEARCH: {
+        return action.data.trim();
```
Comment (neallred) [inline reducers/views/search.ts:29]: I think we'll want to modify this somehow to allow searches for `<firstname> <lastname>`.

https://user-images.githubusercontent.com/13738432/161801424-b390a37a-96a2-44bc-abee-aaaa4a5c623b.mp4

File: reducers/views/search.ts [hunk 6]
```diff
@@ -23,6 +23,19 @@ function modalSearch(state = '', action: GenericAction) {
...
+    }
+
+    case UserTypes.LOGOUT_SUCCESS:
```
Comment (neallred) [inline reducers/views/search.ts:32]: +1 for not accidentally sharing this state between users!

**Review by neallred — COMMENTED** [review]
Overall looks awesome! Found a few corner cases that I think we should address, but otherwise this works great and I love the test coverage.

**Comment by JulienTant** [comment]
> I think we shouldn't show the members info rhs if user is viewing a channel/group message and switches to a direct message. What makes sense to me is:
> 
> * If they have a previousRHSState where they were viewing a different RHS tab, move them to that.
> * If there is no previousRHSState, close the RHS.
> 
>  close-channel-members-when-switch-to-dms.mp4

Should be fixed in the latest push

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Had another look and noticed a couple things beyond what @neallred pointed out with long names:

1. Rather than text wrapping, I'd rather see truncation and to keep things always on one line. We'll definitely need this in manage mode as well for when the dropdown appears.
2. This should be resolved if we truncate long names, but a horizontal scrollbar appears at the bottom of the RHS when a long name is displayed. In addition the long name shows out beyond the bounds of the row item: 
image

image

4. Can we modify the padding on the left/right of the search input field. I think we can cut that down to 8px instead of 16px on each side;
image

5. Also, for the search input, I think we can remove the 'Search members' label from this instance of the component as it's pretty implied by the icon and the context it sits within

**Comment by JulienTant** [comment]
I believe this last update fixed all the long name issues.

**Review by neallred — APPROVED** [review]
Overflows look good now! Nothing else blocking from me. A couple non-blocking naming nits remain.

**Review by dontoisme — APPROVED** [review]

File: components/channel_members_rhs/index.ts [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+import {displayUsername} from 'mattermost-redux/utils/user_utils';
+import {openDirectChannelToUserId} from 'actions/channel_actions';
+import {openModal} from 'actions/views/modals';
+import {closeRightHandSide, goBack} from 'actions/views/rhs';
+
+import {getPreviousRhsState} from '../../selectors/rhs';
+
+import {UserProfile} from 'mattermost-redux/types/users';
+
+import {setChannelMembersRhsSearchTerm} from '../../actions/views/search';
```
Comment (pvev) [inline components/channel_members_rhs/index.ts:32]: nit: I am always more in favor of use absolute paths

File: components/channel_members_rhs/member_list.tsx [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        }
+    }
+`;
+
+export interface Props {
+    className?: string;
+    channel: Channel;
+    members: ChannelMember[];
+    editing: boolean;
+    title?: JSX.Element | null;
```
Comment (pvev) [inline components/channel_members_rhs/member_list.tsx:66]: if is conditional, can't you just rely on the undefined value instead of null?

File: components/channel_members_rhs/search.tsx [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    onInput: (terms: string) => void;
+}
+
+const SearchBar = ({className, terms, onInput}: Props) => {
+    const {formatMessage} = useIntl();
+    return (
+        <div className={className}>
+            <Input
+                value={terms}
+                onInput={(ev) => onInput(ev.currentTarget.value)}
```
Comment (pvev) [inline components/channel_members_rhs/search.tsx:23]: nit: just for naming conventions, call it just e

File: components/channel_members_rhs/search.tsx [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+                        className='style--none'
+                        onClick={() => onInput('')}
+                        aria-label={formatMessage({
+                            id: 'channel_members_rhs.search_bar.aria.cancel_search_button',
+                            defaultMessage: 'cancel members search',
+                        })}
+                    >
+                        <i className={'icon icon-close-circle'}/>
+                    </button>
+                ) : <></>}
```
Comment (pvev) [inline components/channel_members_rhs/search.tsx:36]: I think it is better to set null instead of an empty fragment.

**Review by pvev — APPROVED** [review]
Awesome work @JulienTant :shipit:

File: components/channel_members_rhs/search.tsx (hunk 10 above)
Comment (JulienTant) [reply components/channel_members_rhs/search.tsx]: Thanks for the feedbacks!

**Review by stevemudie — APPROVED** [review]
Looks good @JulienTant!
